### PR TITLE
Add a batch normalizer that has test and train hooks

### PR DESCRIPTION
**Pull request by jrock08**

### CODE REVIEW

File: aiutils/tftools/batch_normalizer.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+This class is designed to simplify adding batch normalization
+(http://arxiv.org/pdf/1502.03167v3.pdf) to your model by
+managing the state variables associated with it.
+
+Important use note:  The function get_assigner() returns
+an op that must be executed to save the updated state.
+A suggested way to do this is to make execution of the
+model optimizer force it, e.g., by:
+
+  update_assignments = tf.group(bn1.get_assigner(),
```
Comment (tomlepaine) [inline aiutils/tftools/batch_normalizer.py]: I almost get this part but not quite 😜  How does it relate to the optimizer?

File: aiutils/tftools/batch_normalizer.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            self.axes = [0, 1, 2]
+        else:
+            raise ValueError('Input tensor must have rank 2 or 4.')
+
+        self.mean = tf.Variable(tf.constant(0.0, shape=[in_dim]),
+                trainable=False)
+        self.variance = tf.Variable(tf.constant(1.0, shape=[in_dim]),
+                trainable=False)
+        with tf.variable_scope(name):
+            self.beta = tf.get_variable('offset',
```
Comment (tomlepaine) [inline aiutils/tftools/batch_normalizer.py]: Can we call these `offset` and `scale` just so I don't have to remember which is beta and gamma?

File: aiutils/tftools/batch_normalizer.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    Use:
+        ewma = tf.train.ExponentialMovingAverage(decay=0.99)
+        bn = BatchNormalizer(input, 0.001, ewma, True)
+        update_assignments = bn.get_assigner()
+        x = bn.normalize(y, train=training?)
+        (the output x will be batch-normalized).
+    """
+
+    def __init__(self, input, epsilon, ewma_trainer, name):
```
Comment (tomlepaine) [inline aiutils/tftools/batch_normalizer.py]: Can we hard cord epsilon, or at least make it have a sensible default?

File: aiutils/tftools/batch_normalizer.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    shape = [in_dim],
+                    initializer=tf.constant_initializer(1.0))
+        self.ewma_trainer = ewma_trainer
+        self.epsilon = epsilon
+
+        # initializes the moving average
+        self.assigner = None
+        self.assigner = self.get_assigner()
+
+    def get_vars(self):
```
Comment (tomlepaine) [inline aiutils/tftools/batch_normalizer.py]: Is this useful?

File: aiutils/tftools/batch_normalizer.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        """Returns an EWMA apply op that must be invoked after optimization."""
+        if not self.assigner:
+            self.assigner = self.ewma_trainer.apply([self.mean, self.variance])
+        return self.assigner
+
+    def normalize(self, input, train=True):
+        """Returns a batch-normalized version of x."""
+        if train:
+            mean, variance = tf.nn.moments(input, self.axes)
+            assign_mean = self.mean.assign(mean)
```
Comment (tomlepaine) [inline aiutils/tftools/batch_normalizer.py]: What do the assignment and control dependency parts here do?

File: aiutils/tftools/batch_normalizer.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                        self.gamma, self.epsilon)
+        else:
+            mean = self.ewma_trainer.average(self.mean)
+            variance = self.ewma_trainer.average(self.variance)
+            local_beta = tf.identity(self.beta)
+            local_gamma = tf.identity(self.gamma)
+            return tf.nn.batch_normalization(input, mean, variance, local_beta,
+                    local_gamma, self.epsilon)
+
+def batch_normalize(input, phase_train, bn_obj):
```
Comment (tomlepaine) [inline aiutils/tftools/batch_normalizer.py]: Can we hide the `bn_obj` inside this function to make it easier to the end-user?

It would be great if the user could just call `batch_normalize(input, phase_train, name)` and have things work.

**Comment by tomlepaine** [comment]
Nice code. But it seems like it requires the user to know things and I wish it didn’t.

Maybe the assignment stuff which I don’t quite understand makes this hard.

I like the class interface, but instead of the user doing:

```
ewma = tf.train.ExponentialMovingAverage(decay=0.99)
bn = BatchNormalizer(input, 0.001, ewma, True)
update_assignments = bn.get_assigner()
x = bn.normalize(y, train=training?)
```

Can the user do:

```
x, update_assignments = batch_normalize(input, phrase_train, name)
```

And `batch_normalize` takes care of everything else?

File: aiutils/tftools/batch_normalizer.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+from tensorflow.python import control_flow_ops
+
+class BatchNormalizer(object):
+    """Helper class that groups the normalization logic and variables.
+
+    Use:
+        ewma = tf.train.ExponentialMovingAverage(decay=0.99)
+        bn = BatchNormalizer(input, 0.001, ewma, True)
+        update_assignments = bn.get_assigner()
+        x = bn.normalize(y, train=training?)
```
Comment (tomlepaine) [inline aiutils/tftools/batch_normalizer.py]: There is a stray `?` on this line.

**Comment by kevjshih** [comment]
Is it usual to include weight decay for bn parameters?

File: aiutils/tftools/batch_normalizer.py (hunk 1 above)
Comment (kevjshih) [inline aiutils/tftools/batch_normalizer.py]: Do you think there's a mistake in this example here? The line `optimizer = tf.group(update_assignments)` doesn't make much sense to me, especially considering this is the second grouping of the updates. Also, I assume we'd just run 
```python 
# skipping quite a bit
loss = create_loss_graph()
train = optimizer.minimize(loss)
with sess.as_default():
     # get the update assignments
     ...
     update_assignments.run()
     train.run(batch_feed_stuff)
```

I know the documentation for the moving average also calls tf.group twice for some reason. Am I missing something?

File: aiutils/tftools/batch_normalizer.py (hunk 5 above)
Comment (kevjshih) [inline aiutils/tftools/batch_normalizer.py]: Assignment creates an op for updating the moving averages. You'll need to run these ops for them to update. 

Here's the documentation for control_dependencies:
https://www.tensorflow.org/versions/master/api_docs/python/framework.html#Graph.control_dependencies

In this example, I think it's saying execute whatever ops are associated with optimizer first, then re-assign the moving average ops to optimizer and then run them. 

It's similar to the example in:
https://www.tensorflow.org/versions/r0.8/api_docs/python/train.html#ExponentialMovingAverage

though I think there may be a typo because overriding the optimizer variable doesn't make sense and doesn't happen in the documented example either.

File: aiutils/tftools/batch_normalizer.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        (the output x will be batch-normalized).
+    """
+
+    def __init__(self, input, epsilon, ewma_trainer, name):
+        rank = len(input.get_shape().as_list())
```
Comment (BigRedT) [inline aiutils/tftools/batch_normalizer.py]: why not `rank = len(input.get_shape())` ? 
I tried it, works fine.

File: aiutils/tftools/batch_normalizer.py (hunk 1 above)
Comment (jrock08) [inline aiutils/tftools/batch_normalizer.py]: So, basically, what this is saying is that you should make it so that the optimizer has to run before the assigners run, then rather than running the optimizer, you should run the assigners, which will automatically run the optimizer.  I think the second group can probably be replaced with an identity.  But I'm not 100% sure (see below)

From the doc https://www.tensorflow.org/versions/r0.8/api_docs/python/framework.html#Graph.control_dependencies

```
# WRONG
def my_func(pred, tensor):
  t = tf.matmul(tensor, tensor)
  with tf.control_dependencies([pred]):
    # The matmul op is created outside the context, so no control
    # dependency will be added.
    return t
# RIGHT
def my_func(pred, tensor):
  with tf.control_dependencies([pred]):
    # The matmul op is created in the context, so a control dependency
    # will be added.
    return tf.matmul(tensor, tensor)
```

File: aiutils/tftools/batch_normalizer.py (hunk 5 above)
Comment (jrock08) [inline aiutils/tftools/batch_normalizer.py]: Yeah, Kevin has the control here right.  See my previous comment above for the explanation on replacing the optimizer.

File: aiutils/tftools/batch_normalizer.py (hunk 6 above)
Comment (jrock08) [inline aiutils/tftools/batch_normalizer.py]: Makes it hard to update the batch norm moving average.  I think keeping around the other batch_normalization code for easy use, might be worth doing.  Alternatively, we do as you suggest, and just hard code it to only do training phase, so it wouldn't update the moving average, but also wouldn't ever use it.  It's probably inefficient to do that since this has many extra ops

File: aiutils/tftools/batch_normalizer.py (hunk 1 above)
Comment (kevjshih) [inline aiutils/tftools/batch_normalizer.py]: Oh, so optimizer is just a reference to the optimizer, and now that it's run by control_dependencies, it's ok to re-assign it to the updater ops?

Comment (jrock08) [inline aiutils/tftools/batch_normalizer.py]: So, basically what it's saying is

```
opt = trainer.minimize(loss)
with tf.control_dependencies([opt]):
  new_opt = tf.group(<all batch norm updates>)
```

So, new_opt is actually the batch norm updates, but it calls the minimizer when you run it.

**Comment by tomlepaine** [comment]
So I'm pretty happy with the final interface, but I wish the class wasn't so big 😿 . I guess we can leave that as an exercise for future motivated collaborators. **TODO**: Shrink the implementation while maintaining the same behavior and interface.

So LGTM 🎉 (once you get the tests to pass).
